packages/nx/src/command-line/migrate.ts

```diff
@@ -12,92 +12,109 @@ import {
+}
+
+export interface ResolvedMigrationConfiguration extends MigrationsJson {
+  packageGroup?: NxMigrationsConfiguration['packageGroup'];
+}
+
+const execAsync = promisify(exec);
+
 export function normalizeVersion(version: string) {
-  const [v, t] = version.split('-');
-  const [major, minor, patch] = v.split('.');
-  const newV = `${major || 0}.${minor || 0}.${patch || 0}`;
-  const newVersion = t ? `${newV}-${t}` : newV;
+  const [semver, prereleaseTag] = version.split('-');
+  const [major, minor, patch] = semver.split('.');
 
-  try {
-    gt(newVersion, '0.0.0');
-    return newVersion;
-  } catch (e) {
+  const newSemver = `${major || 0}.${minor || 0}.${patch || 0}`;
+
+  const newVersion = prereleaseTag
+    ? `${newSemver}-${prereleaseTag}`
+    : newSemver;
+
+  const withoutPatch = `${major || 0}.${minor || 0}.0`;
+  const withoutPatchAndMinor = `${major || 0}.0.0`;
+
+  const variationsToCheck = [
+    newVersion,
+    newSemver,
+    withoutPatch,
+    withoutPatchAndMinor,
+  ];
+
+  for (const variation of variationsToCheck) {
     try {
-      gt(newV, '0.0.0');
-      return newV;
-    } catch (e) {
-      const withoutPatch = `${major || 0}.${minor || 0}.0`;
-      try {
-        if (gt(withoutPatch, '0.0.0')) {
-          return withoutPatch;
-        }
-      } catch (e) {
-        const withoutPatchAndMinor = `${major || 0}.0.0`;
-        try {
-          if (gt(withoutPatchAndMinor, '0.0.0')) {
-            return withoutPatchAndMinor;
-          }
-        } catch (e) {
-          return '0.0.0';
-        }
+      if (gt(variation, '0.0.0')) {
+        return variation;
       }
-    }
+    } catch {}
   }
+
+  return '0.0.0';
 }
 
-function slash(packageName) {
+function normalizeSlashes(packageName: string): string {
   return packageName.replace(/\\/g, '/');
 }
 
+export interface MigratorOptions {
+  packageJson: any;
+  versions: (pkg: string) => string;
+  fetch: (
+    pkg: string,
+    version: string
+  ) => Promise<ResolvedMigrationConfiguration>;
+  to: { [pkg: string]: string };
+}
+
 export class Migrator {
-  private readonly packageJson: any;
-  private readonly versions: (p: string) => string;
-  private readonly fetch: (p: string, v: string) => Promise<MigrationsJson>;
-  private readonly from: { [p: string]: string };
-  private readonly to: { [p: string]: string };
-
-  constructor(opts: {
-    packageJson: any;
-    versions: (p: string) => string;
-    fetch: (p: string, v: string) => Promise<MigrationsJson>;
-    from: { [p: string]: string };
-    to: { [p: string]: string };
-  }) {
+  private readonly packageJson: MigratorOptions['packageJson'];
+  private readonly versions: MigratorOptions['versions'];
+  private readonly fetch: MigratorOptions['fetch'];
+  private readonly to: MigratorOptions['to'];
+
+  constructor(opts: MigratorOptions) {
     this.packageJson = opts.packageJson;
     this.versions = opts.versions;
     this.fetch = opts.fetch;
-    this.from = opts.from;
```

Comment:
Was this meant to be dropped? IIRC it is used when running `nx migrate --from A.B.C --to X.Y.Z`

Reply:
It is only used in the `versions` method, not in `Migrator`'s logic. It was even showing as an unused variable of the class 🤷🏻‍♂️

Reply:
Gotcha, I wasn't sure 👍